pgx/src/datum/numeric.rs

```diff
@@ -7,218 +7,225 @@ All rights reserved.
-        struct NumericVisitor;
-
-        impl<'de> Visitor<'de> for NumericVisitor {
-            type Value = Numeric;
-
-            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-                formatter.write_str("a JSON number or a \"quoted JSON number\"")
-            }
-
-            #[inline]
-            fn visit_i64<E>(self, value: i64) -> Result<Numeric, E> {
-                Ok(value.into())
-            }
-
-            #[inline]
-            fn visit_u64<E>(self, value: u64) -> Result<Numeric, E> {
-                Ok(value.into())
-            }
-
-            #[inline]
-            fn visit_f64<E>(self, value: f64) -> Result<Numeric, E>
-            where
-                E: de::Error,
-            {
-                let result =
-                    Number::from_f64(value).ok_or_else(|| de::Error::custom("not a JSON number"));
-                match result {
-                    Ok(num) => Ok(num.as_f64().unwrap().into()),
-                    Err(e) => Err(e),
-                }
-            }
-
-            #[inline]
-            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
-            where
-                E: Error,
-            {
-                self.visit_string(v.to_owned())
-            }
-
-            #[inline]
-            fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
-            where
-                E: Error,
-            {
-                // try to convert the provided String value into a Postgres Numeric Datum
-                // if it doesn't raise an conversion error, then we're good
-                PgTryBuilder::new(|| {
-                    // this might throw, but that's okay
-                    let datum = Numeric(v.clone()).into_datum().unwrap();
-
-                    unsafe {
-                        // and don't leak the 'inet' datum Postgres created
-                        pg_sys::pfree(datum.cast_mut_ptr());
-                    }
-
-                    // we have it as a valid String
-                    Ok(Numeric(v.clone()))
-                })
-                .catch_when(PgSqlErrorCode::ERRCODE_INVALID_TEXT_REPRESENTATION, |_| {
-                    Err(Error::custom(format!("invalid Numeric value: {}", v)))
-                })
-                .execute()
+impl Drop for AnyNumeric {
+    fn drop(&mut self) {
+        if self.need_pfree {
+            unsafe {
+                pg_sys::pfree(self.inner.cast());
             }
         }
-
-        deserializer.deserialize_any(NumericVisitor)
     }
 }
 
-impl From<i8> for Numeric {
-    fn from(val: i8) -> Self {
-        Numeric(val.to_string())
+impl Display for AnyNumeric {
+    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+        let numeric_out = unsafe {
+            direct_function_call::<&CStr>(pg_sys::numeric_out, vec![self.as_datum()]).unwrap()
+        };
+        let s = numeric_out.to_str().expect("numeric_out is not a valid UTF8 string");
+        fmt.write_str(s)
```

Comment:
nit: It's minor but `s.fmt(fmt)` (or equivalently `fmt.pad(s)`) is better if you have a string around already, since then the size/alignment stuff gets respected.